File: analyzers/tests/SonarAnalyzer.Test/TestCases/InsteadOfAny.EntityFramework.Core.vb
Imports System.Collections.Generic
Imports System.Linq
Imports Microsoft.EntityFrameworkCore

Public Class EntityFrameworkTestcases
    Public Class MyEntity2
        Public Property Id As Integer
    End Class

    Public Class MyDbContext
        Inherits DbContext

        Public Property MyEntities As DbSet(Of MyEntity2)

        Public Sub GetEntities(ByVal dbContext As MyDbContext, ByVal ids As List(Of Integer))
            Dim __ As IQueryable(Of MyEntity2) = dbContext.MyEntities.Where(Function(e) ids.Any(Function(i) e.Id = i)) ' Compliant
            __ = dbContext.MyEntities.Where(Function(e) ids.Any(Function(i) e.Equals(i))) ' Compliant
            __ = dbContext.MyEntities.Where(Function(e) ids.Any(Function(i) e.Id > i)) ' Compliant
            __ = dbContext.MyEntities.Where(Function(e) ids.Any(Function(i) TypeOf e.Id Is i)) ' Error [BC30002]

            Dim iqueryable = dbContext.MyEntities.OrderBy(Function(e) e.Id)
            __ = iqueryable.Where(Function(e) ids.Any(Function(i) e.Id = i)) ' Compliant

        End Sub
    End Class
End Class

File: analyzers/tests/SonarAnalyzer.Test/TestCases/InsteadOfAny.EntityFramework.Core.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

// https://github.com/SonarSource/sonar-dotnet/issues/7286
public class EntityFrameworkReproGH7286
{
    public class MyEntity
    {
        public int Id { get; set; }
    }

    public class FirstContext: DbContext
    {
        public DbSet<MyEntity> MyEntities { get; set; }
    }

    public class SecondContext: DbContext
    {
        public DbSet<MyEntity> SecondEntities { get; set; }
    }


    public void GetEntities(FirstContext dbContext, List<int> ids)
    {
        _ = dbContext.MyEntities.Where(e => ids.Any(i => e.Id == i)); // Compliant
        _ = dbContext.MyEntities.Where(e => ids.Any(i => e.Equals(i))); // Compliant
        _ = dbContext.MyEntities.Where(e => ids.Any(i => e.Id > i)); // Compliant

        _ = dbContext.MyEntities.Where(e => ids.Any(i => e.Id is i)); // Error [CS9135]
        _ = dbContext.MyEntities.Where(e => ids.Any(i => e.Id is 2)); // Error [CS8122]

        var iqueryable = dbContext.MyEntities.OrderBy(e => e.Id);
        _ = iqueryable.Where(e => ids.Any(i => e.Id == i)); // Compliant
    }

    public async Task GetEntitiesAsync(SecondContext secondContext, List<int> ids)
    {
        _ = await secondContext
                .SecondEntities
                .Where(e => ids.Any(i => e.Id == i))
                .ToListAsync();
    }
}
